File: java/validate_subsequence/ValidateSubsequence.java
package main.java.com.example.validatesubsequence;

import java.util.List;

public class ValidateSubsequence {

    /**
     * Determines if the given sequence is a valid subsequence of the given array.
     * A sequence is considered a valid subsequence if all elements of the sequence appear in the array
     * in the same order, but not necessarily consecutively.
     *
     * @param array    The list of integers representing the array.
     * @param sequence The list of integers representing the sequence to be checked.
     * @return true if the sequence is a valid subsequence of the array, false otherwise.
     * 
     */
	// O(n) time | O(1) space; n is the length of the array or the sequence, whichever is greater
    public static boolean isValidSubsequence(List<Integer> array, List<Integer> sequence) {
        int arrIdx = 0; // Index to traverse the array
        int seqIdx = 0; // Index to traverse the sequence
        
        // Loop through the array and sequence
        while (arrIdx < array.size() && seqIdx < sequence.size()) {            
            // If current elements match, move to the next element in the sequence
            if (array.get(arrIdx).equals(sequence.get(seqIdx))) {
                seqIdx++;
            }
            // Always move to the next element in the array
            arrIdx++;
        }
        // If seqIdx has reached the end of the sequence, it means the entire sequence was found
        return seqIdx == sequence.size();
    }   
}

File: python/easy/validate_subsequence/validate_subsequence.py
# O(n) time | O(1) space; n is the length of the array or the sequence, whichever is greater
def isValidSubsequence(array, sequence):
    # Initialize indices for the array and the sequence
    arrIdx = 0
    seqIdx = 0

    # Iterate through the array until we either exhaust the array or find the entire sequence
    while arrIdx < len(array) and seqIdx < len(sequence):
        # Check if the current element in the array matches the current element in the sequence
        if array[arrIdx] == sequence[seqIdx]:
            # Move to the next element in the sequence
            seqIdx += 1
        # Move to the next element in the array
        arrIdx += 1

    # Return True if we have matched all elements in the sequence, otherwise return False
    return seqIdx == len(sequence)